compiler/noirc_evaluator/src/ssa/function_builder/mod.rs

```diff
@@ -259,9 +260,118 @@ impl FunctionBuilder {
+        let typ = self.current_function.dfg.type_of_value(lhs);
+        let pow = self.insert_cast(pow, typ);
+        self.insert_binary(lhs, BinaryOp::Mul, pow)
+    }
+
+    /// Insert ssa instructions which computes lhs << rhs by doing lhs*2^rhs
+    /// and truncate the result to bit_size
+    fn insert_wrapping_shift_left(
+        &mut self,
+        lhs: ValueId,
+        rhs: ValueId,
+        bit_size: u32,
+    ) -> InsertInstructionResult {
+        let base = self.field_constant(FieldElement::from(2_u128));
+        let typ = self.current_function.dfg.type_of_value(lhs);
+        let (max_bit, pow) = if let Some(rhs_constant) =
+            self.current_function.dfg.get_numeric_constant(rhs)
+        {
+            // Happy case is that we know precisely by how many bits the the integer will
+            // increase: lhs_bit_size + rhs
+            let (rhs_bit_size_pow_2, overflows) =
+                2_u32.overflowing_pow(rhs_constant.to_u128() as u32);
+            if overflows {
+                let zero = self.numeric_constant(FieldElement::zero(), typ);
+                return InsertInstructionResult::SimplifiedTo(zero);
+            }
+            let pow = self.numeric_constant(FieldElement::from(rhs_bit_size_pow_2 as u128), typ);
+            (bit_size + (rhs_constant.to_u128() as u32), pow)
+        } else {
+            // we use a predicate to nullify the result in case of overflow
+            let bit_size_var =
+                self.numeric_constant(FieldElement::from(bit_size as u128), typ.clone());
+            let overflow = self.insert_binary(rhs, BinaryOp::Lt, bit_size_var);
+            let one = self.numeric_constant(FieldElement::one(), Type::unsigned(1));
+            let predicate = self.insert_binary(overflow, BinaryOp::Eq, one);
+            let predicate = self.insert_cast(predicate, typ.clone());
+
+            let pow = self.pow(base, rhs);
+            let pow = self.insert_cast(pow, typ);
+            (FieldElement::max_num_bits(), self.insert_binary(predicate, BinaryOp::Mul, pow))
+        };
+
+        let instruction = Instruction::Binary(Binary { lhs, rhs: pow, operator: BinaryOp::Mul });
+        if max_bit <= bit_size {
+            self.insert_instruction(instruction, None)
+        } else {
+            let result = self.insert_instruction(instruction, None).first();
+            self.insert_instruction(
+                Instruction::Truncate { value: result, bit_size, max_bit_size: max_bit },
+                None,
+            )
+        }
+    }
+
+    /// Insert ssa instructions which computes lhs >> rhs by doing lhs/2^rhs
+    pub(crate) fn insert_shift_right(&mut self, lhs: ValueId, rhs: ValueId) -> ValueId {
+        let base = self.field_constant(FieldElement::from(2_u128));
+        let pow = self.pow(base, rhs);
+        self.insert_binary(lhs, BinaryOp::Div, pow)
+    }
+
+    /// Computes lhs^rhs via square&multiply, using the bits decomposition of rhs
```

Comment:
Can you add pseudocode of the algorithm here to the comment? E.g.
```rs
let mut r = 1;
let rhs_bits = to_bits(rhs);
for i in 1 .. bit_size + 1 {
    let r1 = r * r;  // nitpick: should this be r_squared (since r2 is taken)? Alternatively, rename r2 and use that
    let b = rhs_bits[bit_size - i];
    r = (r1 * lhs * b) + (1 - b) * r1;
}
```
Or possibly just a comment of the for loop contents since the rest is more obvious.

Reply:
Done